### PR TITLE
Also treat `impl` definition parent as transparent regarding modules

### PR DESCRIPTION
**Pull request by Urgau**
This PR changes the `non_local_definitions` lint logic to also consider `impl` definition parent as transparent regarding modules.

See tests and explanation in the changes.

@rustbot label +L-non_local_definitions
Fixes *(after beta-backport)* #132427
cc @leighmcculloch
r? @jieyouxu

### CODE REVIEW

**Comment by jieyouxu** [comment]
(Nit: did u mean threat -> thread in PR title and commit msg?)

**Comment by Urgau** [comment]
> (Nit: did u mean threat -> thread in PR title and commit msg?)

Nope, I meant `treat`.

File: tests/ui/lint/non-local-defs/convoluted-locals-131474-with-mods.rs
Comment (jieyouxu) [inline tests/ui/lint/non-local-defs/convoluted-locals-131474-with-mods.rs:67]: Question: ... just to double-check: does this logic also work if there's deeper nesting or mixed interleavings between layers of `mod`s and anon consts `const _: ()`? Like

```rs
mod a {
    mod b {
        const _: () = {
            pub enum ArbitraryContext {}

            const _: () = {
                impl ArbitraryContext {}
            };
        };
    }
}
```

or

```rs
mod foo {
    const _: () = {
        mod auth {
            const _: () = {
                pub enum ArbitraryContext {}

                const _: () = {
                    impl ArbitraryContext {}
                };
            };
        }
    };
}
```

Now, usually one does not simply write such code, but alas.

**Review by jieyouxu — APPROVED** [review]
Thanks, this looks reasonable to me, one question about uhh more "exotic" code patterns.

File: tests/ui/lint/non-local-defs/convoluted-locals-131474-with-mods.rs
Comment (Urgau) [reply tests/ui/lint/non-local-defs/convoluted-locals-131474-with-mods.rs:67]: It should, let me check and add tests for those as well.

Comment (Urgau) [reply tests/ui/lint/non-local-defs/convoluted-locals-131474-with-mods.rs]: And it indeed does. Extra tests added.

**Review by jieyouxu — APPROVED** [review]
Thanks, this LGTM, feel free to r=me after PR CI is green.

**Comment by jieyouxu** [comment]
Nominating for beta-backport: this PR fixes `non_local_definitions` lint to not mistakenly fire on a code pattern such as:

```rs
pub mod auth {
    const _: () = {
        pub enum ArbitraryContext {}

        const _: () = {
            impl ArbitraryContext {}
        };
    };
}
```

where the anon-const was nested inside modules but to the `impl`, we only considered it to be transparent to the anon-const but not the module. This PR fixes that by considering `impl` definition parent as transparent with respect to modules.

@rustbot label +beta-nominated

File: tests/ui/lint/non-local-defs/convoluted-locals-132427.rs
Comment (jieyouxu) [inline tests/ui/lint/non-local-defs/convoluted-locals-132427.rs:12]: Remark (for myself): does this peeling logic work for `macro_generated!` inner impls? I think it should, because it operates based on `DefId`s AFAICT and not spans.

Comment (Urgau) [reply tests/ui/lint/non-local-defs/convoluted-locals-132427.rs:12]: Yes, it does work for macro-generated code, for the reason you explained.
Fro example #132427 was from a proc-macro.

**Review by jieyouxu — APPROVED** [review]

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-11-07/near/481140782). A backport PR will be authored by the release team at the end of the current development cycle.

@rustbot label +beta-accepted

**Pull request merged**
